rspace/src/test/scala/coop/rchain/rspace/history/TrieStructureTests.scala

```diff
@@ -0,0 +1,194 @@
+package coop.rchain.rspace.history
+
+import java.nio.ByteBuffer
+
+import org.lmdbjava.Txn
+import scodec.Codec
+import scodec.bits.ByteVector
+import scodec.codecs._
+
+class TrieStructureTests
+    extends HistoryTestsBase[Txn[ByteBuffer], TestKey, ByteVector]
+    with WithLMDBStore {
+
+  implicit val codecByteVector: Codec[ByteVector] = variableSizeBytesLong(int64, bytes)
+
+  def withTrie[R](f: Trie[TestKey, ByteVector] => R): R =
+    withTestTrieStore { store =>
+      store.withTxn(store.createTxnRead()) { txn =>
+        val trieOpt = store.get(txn, store.workingRootHash.get)
+        trieOpt should not be empty
+        f(trieOpt.get)
+      }
+    }
+
+  def withTrieTxnAndStore[R](
+      f: (ITrieStore[Txn[ByteBuffer], TestKey, ByteVector],
+          Txn[ByteBuffer],
+          Trie[TestKey, ByteVector]) => R): R =
+    withTestTrieStore { store =>
+      store.withTxn(store.createTxnRead()) { txn =>
+        val trieOpt = store.get(txn, store.workingRootHash.get)
+        trieOpt should not be empty
+        f(store, txn, trieOpt.get)
+      }
+    }
+
+  "insert's effect" should "be visible in the outer read transaction" ignore {
+    withTrieTxnAndStore { (store, txn, trie) =>
+      val f = singleElementFixture
+      import f._
+
+      insert(store, key1, val1)
+      // Insert was made in a nested transaction, so it's effect should be visible
+      store.get(txn, store.workingRootHash.get) should not be None
+    }
+  }
+
+  behavior of "A trie"
+  it should "be created as an empty pointer block" in
+    withTrie {
+      case Node(PointerBlock(vector)) =>
+        vector should have size 256
+        vector should contain only None
+      case _ => fail("expected a node")
+    }
+
+  private[this] def singleElementFixture = new {
+    val key1    = TestData.key1
+    val val1    = TestData.val1
+    val rootHex = "0x538b82b41d4360492c17a112864ffb989571504b73b1281677a692e9fd2ee4cc"
+    val leafHex = "0x8d329ed700f130f40b15b73b1bd4f7b70d982acb9dce55e58f58425038f5db1c"
+  }
+
+  private[this] def commonPrefixFixture = new {
+    val key1 = TestData.key1
+    val val1 = TestData.val1
+    val key2 = TestData.key2
+    val val2 = TestData.val2
+
+    val rootHex   = "0xae608338e532497ea0844d9efd6008251451538497d30d3d4dbfb9a032fc8feb"
+    val level1Hex = "0xc67f3d92de6f9e75b9561274fbcc4c8efe8d1161554e3b0f66307a111ddd11ce"
+    val level2Hex = "0xdb61c95e4ea234de4fe154861f86d090af2029da3cfca83cf210c71c539f1942"
+    val level3Hex = "0x681aff745729ccec1d5e3f23f5b56796416a6604a8a07b646fc16f1b0e8c70c8"
+    val leaf1Hex  = "0x8d329ed700f130f40b15b73b1bd4f7b70d982acb9dce55e58f58425038f5db1c"
+    val leaf2Hex  = "0xf22c71982cf8663fb1ea77a444233c99d8c00cd187b0253cfc4213228fea6625"
+
+  }
+
+  it should "have two levels after inserting one element" in {
+    withTestTrieStore { implicit store =>
+      val f = singleElementFixture
+      import f._
+      insert(store, key1, val1)
+
+      assertSingleElementTrie
```

Comment:
I can live with this here in a small context, but it makes me a bit uneasy not to be able to see the actual assertion I'm making directly in the test (e.g. what is the expected value(s) vs actual value(s) or other properties i'm making assertions about).

Reply:
I share your feelings, but there was so much code duplication that I decided to put it in one place (DRY). I can revert that though, I'm not forcing this approach